WalletWasabi.Gui/Global.cs

```diff
@@ -119,87 +119,101 @@ public Global(string dataDir, Config config, UiConfig uiConfig, WalletManager wa
 
 		private TaskCompletionSource<bool> InitializationCompleted { get; } = new();
 
-		private bool InitializationStarted { get; set; } = false;
+		private volatile bool _disposeRequested;
 
-		private CancellationTokenSource StoppingCts { get; }
+		/// <summary>Lock that makes sure the application initialization and dispose methods do not run concurrently.</summary>
+		private AsyncLock InitializationAsyncLock { get; } = new();
+
+		/// <summary>Cancellation token to cancel <see cref="InitializeNoWalletAsync(TerminateService)"/> processing.</summary>
+		private CancellationTokenSource StoppingCts { get; } = new();
 
 		public async Task InitializeNoWalletAsync(TerminateService terminateService)
 		{
-			InitializationStarted = true;
-			var cancel = StoppingCts.Token;
-
-			try
+			using (await InitializationAsyncLock.LockAsync())
```

Comment:
Cancel the lock with StoppingCts?

Reply:
Actually, no, because `StoppingCts` can be disposed and we would get an "DisposedObject" exception.

Reply:
ACK, it would be a very edge case to get into this situation also in the next line we are returning if disposed.

Reply:
I'm able to simulate it and get that exception as can be seen in the OP https://github.com/zkSNACKs/WalletWasabi/pull/6310#issue-727811711

Reply:
This has more consequences. If StoppingCts cancelled the InitializeNoWalletAsync throw OpCancelled. There won't be any ObjectDisposed.

Reply:
The lock with StoppingCts will make this impossible, won't it?

Reply:
Anyway, I see you are confident, it is much reliable than before, let's progress.

Reply:
Well, not having the cancellation token on this line is fine because it will cancel a moment later (most likely). I believe what we have with this PR is safe. We can improve it later but the gains will be inconsequential imho.